[zaidansani] iP

zaidansani opened the pull request:
# zAIbot 
> Pronounced zaɪbɔt. Because who better to name a system about doing stuff other than the person who's doing this instead of something else?

[Latest Release](https://github.com/zaidansani/ip/releases/download/untagged-f08660c4890cd146af81/v0.1.jar)
## Reasons to use this bot

1. I made it 💯 
2. Pretty much it 🤔 

## Features

- [X] add specific tasks
- [X] save tasks to disk, in a human-readable-format
- [X] mark tasks done!
- [ ] GUI
- [ ] take over the world
- [ ] ~~SU this mod~~ (impossible)

Very **cool** `code` _block_:
```Java
public static void main(String[] args) {
     new Zaibot().run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Changed greeting message to match the name of the chatbot

Commit:
Added functionality to echo back command, as well as exit with bye

Commit:
Added the ability to store text entered by user, and display them back when requested.

Commit:
Added ability to mark tasks as done.

Added the Task class.

Commit:
Added support to track ToDos, Deadlines, and Events.

Fixed formatting issues with string messages. Added classes for each task.

Commit:
Added automated UI testing using the I/O redirection technique.

Commit:
Added error handling with incorrect inputs.

Commit:
Added task deletion support.

Commit:
Added the ability to save to file.

Amended test case to work with the new save function as well.

Commit:
Replaced by, from and to fields from String to LocalDateTime

All fields that are supposed to date are stored as String.
Using a DateTime format, or an equivalent would make it more useful.

Let's,
- change the date to be input using a common format YYYY-MM-DD HH:mm
- store the date in the save using a common format (LOCAL-ISO-TIME)

Commit:
Testing: Updated EXPECTED.txt file

EXPECTED.txt file assumed by, from and to fields of events are strings.

Let's
* update the EXPECTED.txt file to display the expected behaviour.

Commit:
Merge branch 'branch-Level-8'

Commit:
TaskList: Create TaskList class

TaskList is stored in Storage, and the string printer in main.
The methods that should be categorised together are not.
This could make the code messier.

Let's:
* create a TaskList class to handle all TaskList related issues
* move all the relevant methods there.

Commit:
Refactor: Move methods to individual classes

Code is mainly centered in the Zaibot file, and there is no separation of responsibilities.

Let's:
* create classes for Command, Parser, Ui, TaskList
* move the relevant methods into their relevant classes
* update the test cases given any change

Commit:
Comments: Add JavaDocs to public methods

Some methods are not commented.
This can cause issues when modifying the codebase in the future.

Let's
* add JavaDocs to all the relevant methods (public facing methods)

Commit:
Comments: Remove wrong comment on main method

Commit:
Refactor: Make Command abstract, replace with subclasses

The Command class handles a lot of separate logic.
The code should be not be mixed.

Let's
* split the commands into four different categories,
namely exit, task addition, listing and update
* update the parser to make commands in this class

Commit:
Refactoring: Made Command abstract, replaced with subclasses

The Command class handles a lot of separate logic.
The code should be not be mixed.

Let's
* split the commands into four different categories,
namely exit, task addition, listing and update
* update the parser to make commands in this class

Commit:
Merge remote-tracking branch 'origin/A-MoreOOP' into A-MoreOOP

Commit:
Merge branch 'A-MoreOOP'

Commit:
Refactor: Divide classes into packages

Classes are all in one folder, without clear organisation.
By putting them in the proper sub-directories and using packaging,
the code will be neater and easier to read.

Let's
* ivide the commands, exceptions, and tasks into their own categories

Commit:
Refactor: Fix accidental edition of strings in previous refactor

Commit:
Merge branch 'branch-A-Packages'

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Test: Add test for Command classes

There is no test cases to check the behaviour of the code.

Let's
* create test cases for all the possible Commands.

Commit:
Merge branch 'branch-A-JUnit'

Commit:
Coding standard: Reformat code

The switch cases are not indented.
This does not follow convention.
Some of the blocks do not follow egyptian brackets.

Let's
* reformat the code to match conventions.

Commit:
Coding standard: Boolean names

The names do not follow the coding standard for boolean.

Let's
* change all boolean methods/fields to match naming conventions.

Commit:
Coding standard: Test names

Test names do not match JUnit conventions.

Let's
* follow JUnit conventions for test names.

Commit:
Remove redundant methods

Some methods which were moved out into subclasses were not removed.
These methods are in the Command class.

Let's remove them.

Commit:
Removed unused imports

Commit:
JavaDoc: Add comments to classes

All classes do not have header comments.

Let's add the comments to the classes.

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Find: Add filter functionality

Listing tasks show all tasks.
There is no functionality to filter by specific input.

Let's:
* add a command to allow filtering
* have the command return the filtered list

Commit:
Test: Add testing for new find command

The new find command has no test cases.

Let's
* add the required test cases.

Commit:
Merge branch 'branch-A-Level9'

Commit:
Merge branch 'master' into branch-A-JavaDoc

Commit:
Add header comment to all classes

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Add GUI to the chatbot

The bot works with a text-UI.

Let's
* add JavaFX to the application
* add the GUI to work with the application
* rework the Zaibot class to work with GUI

Commit:
Update tests to work with refactored methods

Commit:
Remove unnecessary imports

Commit:
Merge branch 'branch-Level-10'

Commit:
Merge branch 'master' into branch-A-CheckStyle

Commit:
Add CheckStyle and fix errors

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Merge branch 'master' into branch-Level-10

Commit:
Add CSS to GUI

Commit:
Merge branch 'branch-Level-10'

Commit:
Separate chat message into lines

Commit:
Merge branch 'branch-Level-10'

Commit:
Fix various issues

* Reformat imports
* Add missing JavaDocs

Commit:
Add suppression filter for CheckStyle

Siddardar commented on src/main/java/zaibot/utils/Storage.java:
```diff
@@ -0,0 +1,139 @@
+        Status status = Status.valueOf(tokens[1].toUpperCase());
+        String name = tokens[2];
+
+        // CHECKSTYLE.OFF: Indentation
+        Task task = switch (tokens[0].trim()) {
+            case "T" -> new ToDoTask(name);
+            case "D" -> new DeadlineTask(name, LocalDateTime.parse(tokens[3]));
+            case "E" ->
+                    new EventTask(name, LocalDateTime.parse(tokens[3]), LocalDateTime.parse(tokens[4]));
+            default ->
+                    throw new ZaibotException("Saved file data not in expected format.");
+        };
+        // CHECKSTYLE.ON: Indentation
```
I think for this mod you have to follow the regular switch statements.

zaidansani replied:
Hi, with regards to this, I think you might want to take a look at this - [a forum topic opened up about the arrowhand notation](https://github.com/nus-cs2103-AY2425S1/forum/issues/49).

Siddardar commented on src/main/java/zaibot/utils/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+    public TaskList filterTasks(String input) {
+        return new TaskList(this.tasks.stream()
+                .filter(task -> task.containsInput(input))
+                .collect(Collectors.toCollection(ArrayList::new)));
```
Damn I didn't even think of using streams definitely makes it super clean. 💯

Siddardar commented on src/test/java/zaibot/command/TaskAdditionCommandTest.java:
```diff
@@ -0,0 +1,146 @@
+    private final TaskList tasks = new TaskList();
+    private final Storage storage = new Storage(tasks);
+
+    @BeforeEach
+    public void setUp() {
+        tasks.clearTasks();
+    }
```
Nice cleanup function

Siddardar reviewed (commented):
Damn bro you're insane. Thanks for giving me ideas on how to improve my own code. If you have time to review my code I'd really appreciate it lol :)

valhrd commented on src/main/java/zaibot/utils/Ui.java:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    private enum Message {
+        GOODBYE("See you whenever."),
+        MARK("Task done. Finally."),
+        UNMARK("Task unmarked. Seriously?"),
+        DELETE("Task removed. Bye bye."),
+        GREETING("Hi, or whatever. What do you want from me today?"),
+        ADD("Another day, another task. Added."),
+        TOTAL("You have %d task(s). Get moving."),
+        FILTER("Filtered the tasks for you. Good enough?");
+
+        private final String msg;
+
+        Message(String msg) {
+            this.msg = msg;
+        }
+
+        @Override
+        public String toString() {
+            return this.msg;
+        }
+    }
```
Nice usage of a private nested class to help organise each message!

valhrd reviewed (commented):
Overall nothing seems wrong, very good code and error handling!

Commit:
Add assertions

Assertions help to check that certain values
are correct (valid states)

Let's add assertions to verify that strings are
non-empty after validation.

Adding assertions liberally is
fine and do not affect the code too much
but guarantee safety to a certain extent.

Commit:
Refactor for code quality

Some names and comments are slightly unclear.

Let's refactor these names and comments to make
them clearer.

Commit:
Replace some loops with streams

Streams are clearer and show the step
by step process of an action better

Let's
* trim all the trailing spaces at once with
a stream
* add to the option map with a stream

This is done by utilising map functions
to trim and collect the options into an array.
Then, we skip the first option and
utilise the map function to generate
the option map.

In addition, as the method parse is long,
we moved out task parsing logic
out into a private method.

zaidansani committed:
Merge pull request #1 from zaidansani/branch-A-Assertions

Add Assertions to the project

Commit:
Merge branch 'master' into branch-A-CodeQuality

Commit:
Merge branch 'master' into branch-A-Streams

zaidansani committed:
Merge pull request #3 from zaidansani/branch-A-CodeQuality

Refactor for code quality

Commit:
Merge branch 'master' into branch-A-Streams

zaidansani committed:
Merge pull request #2 from zaidansani/branch-A-Streams

Add streams

Commit:
Create workflow for Github Actions

Commit:
Add ability to recognise tasks added multiple time

Users can add tasks of the exact same values.

Let's
* add a way to check task equality
* present an error to prevent task duplication

Overriding the equals method helps us check
task equality without explicit typecasting
in the main methods. We can then use this
equals method to create a method
to check if a task already exists.

Commit:
Merge branch 'branch-C-DetectDuplicates'

Commit:
Add user guide

Commit:
Revert "Add user guide"

This reverts commit c44a74f84a2240d3357b46dbe9acd5009665cc6b.

Commit:
Add user guide

Commit:
Fix code style as mentioned in issue

Commit:
Add user guide

Commit:
Add error handling for date input

Commit:
Update bye command

Commit:
Update user guide

Commit:
Add test case for saveString

Commit:
Fix bug related to invalid input
